csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -163,50 +164,122 @@ public static void WriteBool(ref Span<byte> buffer, ref WriterInternalState stat
-                    WriteRawBytes(ref buffer, ref state, bytes);
+                // slowpath when Encoding.GetBytes(Char*, Int32, Byte*, Int32) is not available
+                byte[] bytes = Utf8Encoding.GetBytes(value);
+                WriteRawBytes(ref buffer, ref state, bytes);
 #else
-                    ReadOnlySpan<char> source = value.AsSpan();
-                    int bytesUsed;
-                    unsafe
+                ReadOnlySpan<char> source = value.AsSpan();
+
+                int bytesUsed;
+
+                unsafe
+                {
+                    fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
+                    fixed (byte* destinationBytes = &MemoryMarshal.GetReference(destinationBuffer))
                     {
-                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
-                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer.Slice(state.position)))
-                        {
-                            bytesUsed = Utf8Encoding.GetBytes(sourceChars, source.Length, destinationBytes, buffer.Length);
-                        }
+                        bytesUsed = Encoding.UTF8.GetBytes(sourceChars, source.Length, destinationBytes, destinationBuffer.Length);
                     }
-                    state.position += bytesUsed;
-#endif
                 }
+
+                state.position += bytesUsed;
+#endif
             }
-            else
+        }
+
+        private static void WriteStringMultiBuffer(ref Span<byte> buffer, ref WriterInternalState state, string value)
+        {
+#if NETSTANDARD1_1
+            // slowpath when Encoding.GetBytes(Char*, Int32, Byte*, Int32) is not available
+            byte[] bytes = Utf8Encoding.GetBytes(value);
+            WriteRawBytes(ref buffer, ref state, bytes);
+#else
+            if (state.CodedOutputStream != null)
             {
-                // Opportunity for future optimization:
-                // Large strings that don't fit into the current buffer segment
-                // can probably be optimized by using Utf8Encoding.GetEncoder()
-                // but more benchmarks would need to be added as evidence.
+                // When writing to Encoder we need at least 4 bytes available in the buffer to ensure
+                // there is space to write a character. Because CodedOutputStream always writes to the
+                // end of its buffer we can't guarantee that so use slow path.
                 byte[] bytes = Utf8Encoding.GetBytes(value);
                 WriteRawBytes(ref buffer, ref state, bytes);
+                return;
+            }
+
+            // The destination byte array is not be large enough so multiple writes are required.
+            // Encoder will keep state of unwritten data.
+            if (state.stringEncoder == null)
+            {
+                state.stringEncoder = Encoding.UTF8.GetEncoder();
```

Comment:
same problem with caching the stringEncoder as in the other PR. Not seeing stringEncoder.Reset() anywhere seems suspicious.

Reply:
I've changed it so that the convert loop will stop when `completed` is true. That ensures there is no remaining state.

See remarks - https://docs.microsoft.com/en-us/dotnet/api/system.text.encoder.convert?view=netcore-3.1#System_Text_Encoder_Convert_System_Char__System_Int32_System_Byte__System_Int32_System_Boolean_System_Int32__System_Int32__System_Boolean__

Reply:
I think it would be safer to just call Reset() proactively because it would make the reasoning much easier.
Like this if an exception is thrown, the internal state of encoder will be maintained. I know normally throwing would invalidate the entire write context, so this should be fine in theory, but it's hard to predict all the possible patterns of use so I think relying on this is fragile.
How much more expensive would be to call Reset() proactively each time? (sounds like it should be cheap and if done so there would be absolutely no question if the encoder could be polluted or not).

Reply:
> Like this if an exception is thrown, the internal state of encoder will be maintained.

Is CodedOutputStream suppose to be safe to use after it throws an exception? We could put it in a try/catch and reset on an error, but it doesn't seem useful if other state (buffer content, position, written output, etc) is in invalid states.

> How much more expensive would be to call Reset() proactively each time? (sounds like it should be cheap and if done so there would be absolutely no question if the encoder could be polluted or not).

I can test, but the `completed` result says that the internal buffer has been emptied so Reset would never do anything.

https://docs.microsoft.com/en-us/dotnet/api/system.text.encoder.convert?view=netcore-3.1#System_Text_Encoder_Convert_System_Char__System_Int32_System_Byte__System_Int32_System_Boolean_System_Int32__System_Int32__System_Boolean

> The `completed` output parameter indicates whether all the data in the input buffer was converted and stored in the output buffer.